README.md
# Text

![license](https://img.shields.io/badge/license-MIT-brightGreen.svg)
[![build](https://github.com/originphp/text/workflows/CI/badge.svg)](https://github.com/originphp/text/actions)
[![coverage](https://coveralls.io/repos/github/originphp/text/badge.svg?branch=master)](https://coveralls.io/github/originphp/text?branch=master)

The Text utility has a number of methods to help when working with strings.

## Installation

To install this package

```linux
$ composer require originphp/text
```

## Converting Strings to Ascii

To convert a string into Ascii (Transliterate)

```php
$ascii = Text::toAscii('Ragnarr Loðbrók'); // Ragnarr Lodbrok
```

## Creating a Slug

To create a URL safe slug. The string will be converted to ASCII then any unkown characters will be replaced with the separator.

```php
$slug = Text::slug('Who is Ragnarr Loðbrók?'); // who-is-ragnarr-lodbrok
```

## Contains

To check if a string contains a substring.

```php
$result = Text::contains('foo','What is foo bar'); // true
```

## Getting parts of strings

When you need to get part of a string before or after a substring

```php
$result = Text::left('foo','What is foo bar'); // 'What is '
$result = Text::right('foo','What is foo bar'); //' bar'
```

## Checking the start and end of a string

```php
$bool = Text::startsWith('What','What is foo bar'); // true
$bool = Text::endsWith('bar','What is foo bar'); // true
```

## Replace

To replace a substring with another string

```php
$result = Text::replace('foo','***','What is foo bar'); // 'What is *** bar'
$result = Text::replace('foo','***','What is FOO bar',['insensitive'=>true]); // 'What is *** bar'
```

## Insert

To insert values into a string using placeholders (string interpolation)

```php
$string = Text::insert('Record {id} has been updated',[
    'id'=>1234568
]); // Record 1234568 has been updated
```

Another example:

```php
$letter = file_get_contents('/directory/some-file');
$string = Text::insert($letter,[
    'salutation' => 'Mr.',
    'first_name' => 'Tony',
    'last_name' => 'Robbins',
    'address_1' => '100 Santa Monica Road',
]);
```

You can also change the place holders

```php
$string = Text::insert('Record :id has been updated',[
    'id'=>1234568,'before'=>':','after'=>''
    ]); // Record 1234568 has been updated
```


## Tokenize

For quick and easy parsing of strings, the Tokenize method makes things simple. By default `tokenize` splits strings
using a comma `,` and quotation mark `"` as an enclosure.

```php
$string = '2019-07-10 13:30:00 192.168.1.22 "GET /users/login HTTP/1.0" 200 1024';
$result = Text::tokenize($string,['separator'=>' ']);

/*
// Will give you this
[
    '2019-07-10',
    '13:30:00',
    '192.168.1.22',
    'GET /users/login HTTP/1.0',
    '200',
    '1024'
];
*/
```

You can also supply keys instead which will be mapped.

```php
$string = '2019-07-10 13:30:00 192.168.1.22 "GET /users/login HTTP/1.0" 200 1024';
$result = Text::tokenize($string,[
    'separator'=>' ',
    'keys'=>['date','time','ip','request','code','bytes']
]);

/*
// Will give you this
[
    'date'=>'2019-07-10',
    'time'=>'13:30:00',
    'ip' => '192.168.1.22',
    'request' =>'GET /users/login HTTP/1.0',
    'code'=>'200',
    'bytes'=>'1024'
];
*/
```

## Truncate

To truncate a string if it is longer than a specific length. The default length is 30.

```php
$truncated = Text::truncate($string,['length'=>50,'end'=>'... truncated']);
```

## Word Wrap

To wordwrap a string

```php
$wrapped = Text::wordWrap($string); // default is 80
$wrapped = Text::wordWrap($string,['width'=>50]);
```

## Other

Other handy string functions  (through `mb_string`)

```php
$lowerCase = Text::toLower($string);
$uppserCase = Text::toUpper($string);
$int = Text::length($string);
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.1.0] - 2021-01-19

### Added

- Added Text::toUpper
- Added Text::toLower

### Deprecated

- Deprecated Text::upper use Text::toUpper
- Deprecated Text::lower use Text::toLower

### Fixed

- Fixed operators
- Migrated PHPUnit xml schema

## [2.0.0] - 2021-01-04

### Changed

- Changed minimum PHP version to 7.3
- Change minimum PHPUnit to 9.2

## [1.0.1] - 2019-10-25

### Fixed

- Slug replaces characters with separator instead of removing and trims duplicates

## [1.0.0] - 2019-10-12

This component has been decoupled from the [OriginPHP framework](https://www.originphp.com/).

tests/TextTest.php
<?php

/**
 * OriginPHP Framework
 * Copyright 2018 - 2021 Jamiel Sharief.
 *
 * Licensed under The MIT License
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * @copyright   Copyright (c) Jamiel Sharief
 * @link        https://www.originphp.com
 * @license     https://opensource.org/licenses/mit-license.php MIT License
 */
namespace Origin\Test\Text;

use Origin\Text\Text;
use Exception;

class TextTest extends \PHPUnit\Framework\TestCase
{
    public function testToAscii()
    {
        $this->assertEquals('Ragnarr Lodbrok', Text::toAscii('Ragnarr Loðbrók'));
    }
    public function testSlug()
    {
        $this->assertEquals('who-is-ragnarr-lodbrok', Text::slug('Who is Ragnarr Loðbrók?'));
        $this->assertEquals('v2-0-0-with-extra-spacing', Text::slug('v2.0.0 WITH     extra spacing!'));
        $this->assertEquals('test-trimming', Text::slug('^test||trimming!'));
    }

    public function testTokenize()
    {
        $text = '2019-07-10 13:30:00 192.168.1.22 "GET /users/login HTTP/1.0" 200 1024';
        $expected = [
            '2019-07-10',
            '13:30:00',
            '192.168.1.22',
            'GET /users/login HTTP/1.0',
            '200',
            '1024',
        ];
        $this->assertEquals($expected, Text::tokenize($text));

        $expected = [
            'date' => '2019-07-10',
            'time' => '13:30:00',
            'ip' => '192.168.1.22',
            'request' => 'GET /users/login HTTP/1.0',
            'code' => '200',
            'bytes' => '1024',
        ];
        $this->assertEquals($expected, Text::tokenize($text, ['keys' => ['date','time','ip','request','code','bytes']]));

        // invalid amount of keys
        $this->expectException(Exception::class);
        Text::tokenize($text, ['keys' => ['date']]);
    }

    public function testStartsWith()
    {
        $this->assertTrue(Text::startsWith('foo', 'foobar'));
        $this->assertFalse(Text::startsWith('foo', 'barfoo'));
        $this->assertFalse(Text::startsWith('', 'foobar'));
    }
    public function testEndsWith()
    {
        $this->assertFalse(Text::endsWith('foo', 'foobar'));
        $this->assertTrue(Text::endsWith('foo', 'barfoo'));
        $this->assertFalse(Text::endsWith('', 'foobar'));
    }
    public function testLeft()
    {
        $this->assertEquals('foo', Text::left(':', 'foo:bar'));
        $this->assertEquals('one', Text::left(':', 'one:two:three'));
        $this->assertNull(Text::left('x', 'foo:bar'));
        $this->assertNull(Text::left('', ''));
    }
    public function testRight()
    {
        $this->assertEquals('bar', Text::right(':', 'foo:bar'));
        $this->assertNull(Text::right('x', 'foo:bar'));
        $this->assertNull(Text::right('', ''));
    }
    public function testContains()
    {
        $this->assertTrue(Text::contains('foo', 'foobar'));
        $this->assertTrue(Text::contains('foo', 'barfoo'));
        $this->assertTrue(Text::contains('foo', 'xfoox'));
        $this->assertFalse(Text::contains('moo', 'barfoo'));
        $this->assertFalse(Text::contains('', 'barfoo'));
    }
    public function testUpLo()
    {
        $this->assertEquals(strtoupper('foo'), Text::upper('foo'));
        $this->assertEquals(strtolower('FOO'), Text::lower('FOO'));

        $this->assertEquals(strtoupper('foo'), Text::toUpper('foo'));
        $this->assertEquals(strtolower('FOO'), Text::toLower('FOO'));
    }
    public function testReplace()
    {
        $this->assertEquals('foo', Text::replace('bar', '', 'foobar'));
        $this->assertEquals('foo', Text::replace('bar', '', 'fooBAR', ['insensitive' => true]));
    }
    public function testLen()
    {
        $this->assertEquals(3, Text::length('foo'));
        $var = null;
        $this->assertEquals(0, Text::length($var));
    }

    public function testInsert()
    {
        $this->assertEquals(
            'Record 1234568 has been updated',
            Text::insert('Record {id} has been updated', ['id' => 1234568])
        );

        $this->assertEquals(
            'Record 1234568 has been updated',
            Text::insert('Record {id} has been updated', ['id' => 1234568,['before' => ':','after' => '']])
        );
    }

    /**
     * text from ruby docs
     * @source https://api.rubyonrails.org/classes/ActionView/Helpers/TextHelper.html#method-i-word_wrap
     * @return void
     */
    public function testWordWrap()
    {
        $string = 'Once upon a time';
        $expected = 'Once upon a time';
        $this->assertEquals($expected, Text::wordWrap($string));

        $string = 'Once upon a time, in a kingdom called Far Far Away, a king fell ill, and finding a successor to the throne turned out to be more trouble than anyone could have imagined...';
        $expected = "Once upon a time, in a kingdom called Far Far Away, a king fell ill, and finding\na successor to the throne turned out to be more trouble than anyone could have\nimagined...";
        $this->assertEquals($expected, Text::wordWrap($string));

        $string = 'Once upon a time';
        $expected = "Once\nupon a\ntime";
        $this->assertEquals($expected, Text::wordWrap($string, ['width' => 8]));

        $expected = "Once\nupon\na\ntime";
        $this->assertEquals($expected, Text::wordWrap($string, ['width' => 1]));
    }

    public function testTruncate()
    {
        $string = <<< EOF
Ragnar Lodbrok or Lothbrok (Old Norse: Ragnarr Loðbrók, "Ragnar shaggy breeches", contemporary Norse: Ragnar Loðbrók) was a historically dubious[1][2] Norse Viking hero and legendary king of Denmark and Sweden, known from Viking Age Old Norse poetry and sagas. According to that traditional literature, Ragnar distinguished himself by many raids against Francia and Anglo-Saxon England during the 9th century. There is no evidence that he existed under this name and outside of the mythology associated with him. According to the Tale of Ragnar Lodbrok, Ragnar was the son of the Swedish king Sigurd Ring.
EOF;
        $this->assertEquals('Ragnar Lodbrok or Lothbrok (Ol...', Text::truncate($string));
        $this->assertEquals('Ragnar Lodbrok or Lothbrok', Text::truncate('Ragnar Lodbrok or Lothbrok'));
    }
}


src/Text.php
<?php

/**
 * OriginPHP Framework
 * Copyright 2018 - 2021 Jamiel Sharief.
 *
 * Licensed under The MIT License
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * @copyright   Copyright (c) Jamiel Sharief
 * @link        https://www.originphp.com
 * @license     https://opensource.org/licenses/mit-license.php MIT License
 */
declare(strict_types = 1);
namespace Origin\Text;

use Exception;

/**
 * @see https://api.rubyonrails.org/classes/ActionView/Helpers/TextHelper.html
 */
class Text
{
    /**
    * Checks if a string contains a substring
    *
    * @param string $needle
    * @param string $haystack
    * @return bool
    */
    public static function contains(string $needle, string $haystack) : bool
    {
        return $needle !== '' && (mb_strpos($haystack, $needle) !== false);
    }

    /**
     * Gets part of the string from the left part of characters
     *
     * @param string $characters   :
     * @param string $string    key:value
     * @return string|null      key
     */
    public static function left(string $characters, string $string) : ?string
    {
        if (! empty($characters)) {
            $position = mb_strpos($string, $characters);
            if ($position === false) {
                return null;
            }
    
            return mb_substr($string, 0, $position);
        }

        return null;
    }

    /**
     * Gets part of the string from the right part of characters
     *
     * @param string $characters   :
     * @param string $string    key:value
     * @return string|null     value
     */
    public static function right(string $characters, string $string) : ?string
    {
        if (! empty($characters)) {
            $position = mb_strpos($string, $characters);
            if ($position === false) {
                return null;
            }
    
            return mb_substr($string, $position + mb_strlen($characters));
        }

        return null;
    }

    /**
     * Checks if a string starts with another string
     *
     * @param string $needle
     * @param string $haystack
     * @return boolean
     */
    public static function startsWith(string $needle, string $haystack) : bool
    {
        $length = mb_strlen($needle);

        return  ($needle !== '' && mb_substr($haystack, 0, $length) === $needle);
    }

    /**
     * Checks if a string ends with another string
     *
     * @param string $needle
     * @param string $haystack
     * @return boolean
     */
    public static function endsWith(string $needle, string $haystack) : bool
    {
        $length = mb_strlen($needle);

        return ($needle !== '' && mb_substr($haystack, -$length, $length) === $needle);
    }

    /**
     * Inserts values into a string by replacing placeholders. (string interpolation)
     *
     * @param string $string
     * @param array $values
     * @param array $options (before,after) e.g. ['before'=>':', 'after'=>'']
     * @return string
     */
    public static function insert(string $string, array $values = [], array $options = []) : string
    {
        $options += ['before' => '{','after' => '}'];
        $replace = [];
        foreach ($values as $key => $value) {
            $key = "{$options['before']}{$key}{$options['after']}";
            $replace[$key] = $value;
        }

        return strtr($string, $replace);
    }

    /**
     * Replaces text in strings.
     *
     * @internal str_replace works with multibyte string see https://php.net/manual/en/ref.mbstring.php#109937
     * @param mixed $needle
     * @param mixed $with
     * @param mixed $haystack
     * @param array $options (insensitive = false)
     *  - insensitive: default false. case-insensitive replace
     * @return string
     */
    public static function replace($needle, $with, $haystack, array $options = []) : string
    {
        $options += ['insensitive' => false];
        if ($options['insensitive']) {
            return str_ireplace($needle, $with, $haystack);
        }

        return str_replace($needle, $with, $haystack);
    }

    /**
     * Returns the length of a string
     *
     * @param string $string
     * @return integer
     */
    public static function length(string $string = null) : int
    {
        return mb_strlen((string) $string);
    }

    /**
     * @deprecated use toLower
     * @param string $string
     * @return string
     */
    public static function lower(string $string) :string
    {
        return static::toLower($string);
    }

    /**
     * @deprecated use toUpper
     * @param string $string
     * @return string
     */
    public static function upper(string $string) :string
    {
        return static::toUpper($string);
    }

    /**
     * Alias for transliterate
     *
     * @param string $string
     * @param string $transliterator
     * @return string
     */
    public static function toAscii(string $string, string $transliterator = null) : string
    {
        return static::transliterate($string, $transliterator);
    }

    /**
     * Converts all characters in a string to ASCII
     *
     * @param string $string
     * @param string $transliterator
     * @return string
     * @see https://www.php.net/manual/en/transliterator.transliterate.php
     */
    public static function transliterate(string $string, string $transliterator = null) : string
    {
        if ($transliterator === null) {
            $transliterator = 'Any-Latin; Latin-ASCII; [\u0080-\u7fff] remove';
        }

        $result = transliterator_transliterate($transliterator, $string);
        if ($result) {
            $string = $result;
        }

        return $string;
    }

    /**
     * Parses a string into an array
     *
     * @param string $string
     * @param array $options
     *  - keys: an array of names to be mapped as key names
     *  - separator: default:space
     *  - enclosure: default:"
     *  - escape: default:\ one character only
     * @return array
     */
    public static function tokenize(string $string, array $options = []) : array
    {
        $options += ['keys' => [],'separator' => ' ','enclosure' => '"','escape' => '\\'];
        $result = str_getcsv($string, $options['separator'], $options['enclosure'], $options['escape']);
        if (empty($options['keys'])) {
            return $result;
        }
        $out = [];
        if (count($options['keys']) !== count($result)) {
            throw new Exception('Invalid amount of keys');
        }
        foreach ($options['keys'] as $i => $key) {
            $out[$key] = $result[$i];
        }

        return $out;
    }

    /**
     * Creates a slug for use in a URL the string is converted into ascii then all unmatched
     * characters are changed to
     *
     * @param string $string
     * @param array $options The available options keys are :
     *   - separator: default:-
     * @return string
     */
    public static function slug(string $string, array $options = []) : string
    {
        $options += ['separator' => '-'];
        $ascii = static::transliterate($string);
    
        $string = preg_replace('/[^a-z0-9-]/i', $options['separator'], mb_strtolower($ascii));

        return trim(preg_replace('/-+/', $options['separator'], $string), '-');
    }

    /**
     * Truncates a string
     *
     * @param string $string
     * @param array $options (length:30, end:...)
     * @return string
     */
    public static function truncate(string $string, array $options = []) : string
    {
        $options += ['length' => 30,'end' => '...'];
        if (mb_strwidth($string) <= $options['length']) {
            return $string;
        }
        $string = mb_strimwidth($string, 0, $options['length'], '');

        return rtrim($string) . $options['end'];
    }

    /**
     * Wraps a string to a given number of characters
     *
     * @param string $string
     * @param array $options The options are
     *  - width: default:80
     *  - break: default:\n what to break lines with
     *  - cut: default:false to cut string
     * @return string
     */
    public static function wordWrap(string $string, array $options = []) : string
    {
        $options += ['width' => 80,'break' => "\n",'cut' => false];

        return wordwrap($string, $options['width'], $options['break'], $options['cut']);
    }

    /**
     * Converts a string to lower case
     *
     * @param string $string
     * @return string
     */
    public static function toLower(string $string) :string
    {
        return mb_strtolower($string);
    }

    /**
     * Converts a stirng to uppercase
     *
     * @param string $string
     * @return string
     */
    public static function toUpper(string $string) :string
    {
        return mb_strtoupper($string);
    }
}
